// @bootstrap-vue-next/nuxt/dist/module.mjs
import { addComponent, defineNuxtModule, createResolver, addPlugin, addImports } from '@nuxt/kit';
import { Components, Directives, Composables } from 'bootstrap-vue-next';

const useComponents = () => {
  Object.keys(Components).forEach((name) => {
    addComponent({
      name,
      export: name,
      filePath: "bootstrap-vue-next"
    });
  });
};

const parseActiveImports = (options, values) => {
  const { all, ...others } = options;
  const valuesCopy = {};
  if (all) {
    values.forEach((el) => {
      valuesCopy[el] = all;
    });
  }
  const merge = { ...valuesCopy, ...others };
  return Object.entries(merge).filter(([name, value]) => !!value && values.includes(name)).map(([name]) => name);
};

const normalizeConfigurationValue = (option) => {
  return typeof option === "boolean" ? { all: option } : option;
};

const module = defineNuxtModule({
  meta: {
    name: "bootstrap-vue-next",
    configKey: "bootstrapVueNext",
    compatibility: {
      nuxt: "^3.0.0",
      bridge: false
    }
  },
  defaults: {
    composables: true,
    directives: true
  },
  setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url);
    nuxt.options.build.transpile.push(resolve("./runtime"));
    nuxt.options.css.push("bootstrap-vue-next/dist/bootstrap-vue-next.css");
    const normalizedComposableOptions = normalizeConfigurationValue(options.composables);
    const normalizedDirectiveOptions = normalizeConfigurationValue(options.directives);
    nuxt.options.css.push("bootstrap-vue-next/dist/bootstrap-vue-next.css");
    nuxt.options.vite.optimizeDeps = nuxt.options.vite.optimizeDeps || {};
    nuxt.options.vite.optimizeDeps.include = nuxt.options.vite.optimizeDeps.include || [];
    nuxt.options.vite.optimizeDeps.include.push("bootstrap-vue-next");
    addPlugin(resolve("./runtime/createBootstrap"));
    const transformAssetUrls = Object.freeze({
      BImg: ["src"]
    });
    nuxt.options.vite.vue = nuxt.options.vite.vue || {};
    nuxt.options.vite.vue.template = nuxt.options.vite.vue.template || {};
    nuxt.options.vite.vue.template.transformAssetUrls = nuxt.options.vite.vue.template.transformAssetUrls ?? {};
    if (typeof nuxt.options.vite.vue.template.transformAssetUrls !== "boolean" && !("BImg" in nuxt.options.vite.vue.template.transformAssetUrls || "b-img" in nuxt.options.vite.vue.template.transformAssetUrls)) {
      nuxt.options.vite.vue.template.transformAssetUrls = {
        ...nuxt.options.vite.vue.template.transformAssetUrls,
        ...transformAssetUrls
      };
    }
    useComponents();
    if (Object.values(normalizedDirectiveOptions).some((el) => el === true)) {
      const activeDirectives = parseActiveImports(
        normalizedDirectiveOptions,
        Object.keys(Directives)
      );
      nuxt.options.runtimeConfig.public.bootstrapVueNext = {
        directives: activeDirectives
      };
      addPlugin(resolve("./runtime/useDirectives"));
    }
    if (Object.values(normalizedComposableOptions).some((el) => el === true)) {
      parseActiveImports(normalizedComposableOptions, Object.keys(Composables)).forEach(
        (name) => addImports({
          from: "bootstrap-vue-next",
          name
        })
      );
    }
  }
});

export { module as default };
