// nuxt-simple-robots/dist/module.mjs
import { defineNuxtModule, createResolver, useLogger, addTemplate, addImports, addComponent, addServerHandler } from '@nuxt/kit';
import defu from 'defu';

function asArray(v) {
  return typeof v === "undefined" ? [] : Array.isArray(v) ? v : [v];
}

const module = defineNuxtModule({
  meta: {
    name: "nuxt-simple-robots",
    compatibility: {
      nuxt: "^3.3.1",
      bridge: false
    },
    configKey: "robots"
  },
  defaults(nuxt) {
    let indexable = true;
    if (typeof process.env.NUXT_INDEXABLE !== "undefined")
      indexable = String(process.env.NUXT_INDEXABLE) !== "false";
    else if (typeof nuxt.options.runtimeConfig.indexable !== "undefined")
      indexable = String(nuxt.options.runtimeConfig.indexable) !== "false";
    else if (process.env.NODE_ENV !== "production")
      indexable = false;
    const siteUrl = process.env.NUXT_PUBLIC_SITE_URL || process.env.NUXT_SITE_URL || nuxt.options.runtimeConfig.public?.siteUrl || nuxt.options.runtimeConfig.siteUrl;
    return {
      enabled: true,
      siteUrl,
      disallow: [],
      sitemap: [],
      indexable,
      robotsEnabledValue: "index, follow, max-image-preview:large, max-snippet:-1, max-video-preview:-1",
      robotsDisabledValue: "noindex, nofollow",
      disallowNonIndexableRoutes: false
    };
  },
  async setup(config, nuxt) {
    if (config.enabled === false)
      return;
    config.siteUrl = config.siteUrl || config.host;
    const { resolve } = createResolver(import.meta.url);
    config.indexable = String(config.indexable) !== "false";
    const logger = useLogger("nuxt-simple-robots");
    nuxt.hook("modules:done", async () => {
      config.sitemap = asArray(config.sitemap);
      config.disallow = asArray(config.disallow);
      await nuxt.hooks.callHook("robots:config", config);
      config.sitemap = !config.indexable ? [] : [...new Set(config.sitemap)];
      nuxt.options.routeRules = nuxt.options.routeRules || {};
      Object.entries(nuxt.options.routeRules).forEach(([route, rules]) => {
        if (rules.index === false || rules.robots) {
          nuxt.options.routeRules[route] = defu({
            headers: {
              "X-Robots-Tag": rules.index === false ? config.robotsDisabledValue : rules.robots
            }
          }, nuxt.options.routeRules?.[route]);
        }
      });
      let disallow = config.disallow;
      if (config.disallowNonIndexableRoutes && config.indexable) {
        Object.entries(nuxt.options.routeRules || {}).forEach(([route, rules]) => {
          if (rules.index === false || rules.robots?.includes("noindex")) {
            disallow.push(route.replaceAll("**", "*"));
          }
        });
      }
      if (!config.indexable)
        disallow = ["/"];
      else if (!disallow.length)
        disallow.push("");
      config.disallow = [...new Set(disallow)];
      const hasRelativeSitemaps = config.sitemap.some((sitemap) => !sitemap.startsWith("http"));
      if (hasRelativeSitemaps && !config.siteUrl && nuxt.options._generate)
        logger.warn("You are prerendering your robots.txt but have not set a siteUrl. This will result in relative sitemap URLs which are not valid.");
      nuxt.options.runtimeConfig.public["nuxt-simple-robots"] = config;
    });
    addTemplate({
      filename: "nuxt-simple-robots.d.ts",
      getContents: () => {
        return `// Generated by nuxt-simple-robots
interface NuxtSimpleRobotsNitroRules {
  index?: boolean
  robots?: string
}
declare module 'nitropack' {
  interface NitroRouteRules extends NuxtSimpleRobotsNitroRules {}
  interface NitroRouteConfig extends NuxtSimpleRobotsNitroRules {}
}
export {}
`;
      }
    });
    nuxt.hooks.hook("prepare:types", ({ references }) => {
      references.push({ path: resolve(nuxt.options.buildDir, "nuxt-simple-robots.d.ts") });
    });
    if (nuxt.options._generate) {
      nuxt.hooks.hook("nitro:init", async (nitro) => {
        nitro.options.prerender.routes = nitro.options.prerender.routes || [];
        nitro.options.prerender.routes.push("/robots.txt");
      });
    }
    nuxt.options.optimization.treeShake.composables.client["nuxt-simple-robots"] = ["defineRobotMeta"];
    addImports({
      name: "defineRobotMeta",
      from: resolve("./runtime/composables/defineRobotMeta")
    });
    await addComponent({
      name: "RobotMeta",
      filePath: resolve("./runtime/components/RobotMeta")
    });
    addServerHandler({
      route: "/robots.txt",
      handler: resolve("./runtime/server/robots-txt")
    });
    addServerHandler({
      handler: resolve("./runtime/server/robots-middleware")
    });
  }
});

export { module as default };
